### ISSUE TITLE
AJAX基础--跨域JSONP

### ISSUE THREAD

**Issue by liSNS** [liSNS/Level-4CSS3.HTML5.Expression.Ajax.JQ#4]
<div class="toc"> <ul> <li><a href="#ajax基础瀑布流项目开发留言本实战开发">AJAX基础、瀑布流项目开发、留言本实战开发</a><ul> <li><a href="#ajax基础">AJAX基础</a><ul> <li><a href="#ajax原理">Ajax原理</a></li> <li><a href="#ajax过程详解">Ajax过程详解</a></li> <li><a href="#编写ajax">编写Ajax</a></li> <li><a href="#实例ajax的封装">实例：AJAX的封装</a></li> </ul> </li> <li><a href="#瀑布流项目开发">瀑布流项目开发</a></li> <li><a href="#ajax-跨域解决之jsonp">AJAX-跨域解决之JSONP</a></li> </ul> </li> </ul> </div>
<h2 id="ajax基础">AJAX基础</h2>
<h3 id="ajax原理">Ajax原理</h3>
<p><strong>什么是Ajax</strong></p>
<ul> <li>Ajax：Asynchronous JavaScript and XML（异步的javascript和XML）:用javascript异步的形式去操作xml <br> <ul><li>作用：数据交互</li> <li>节省用户操作时间，提高用户体验，减少数据请求</li> <li>传输获取数据</li></ul></li> </ul>
<p><strong>使用Ajax</strong></p>
<ul>
<li>使用ajax获取某一文本文件的内容</li>
</ul>
<h3 id="ajax过程详解">Ajax过程详解</h3>
<ol> <li>创建一个ajax对象 <br> <ul><li>ie6以上和标准浏览器：new XMLHttpRequest()</li> <li>ie6及以下：new ActiveXObject('Microsoft.XMLHTTP')</li></ul></li> <li>open方法 <br> <ul><li>三个参数： <br> <ul><li>提交方式，相当于form-method</li> <li>提交地址，相当于form-action</li> <li>异步（同步）</li></ul></li></ul></li> <li>send方法 <br> <ul><li>发送数据请求，相当于form的submit</li></ul></li> <li>请求状态监控 <br> <ul><li>onreadystatechange事件 <br> <ul><li>readyState属性：请求状态 <br> <ul><li>0（初始化）还没有调用open()方法</li> <li>1（载入）已调用send()方法，正在发送请求</li> <li>2（载入完成）send()方法完成，已收到全部响应内容</li> <li>3（解析）正在解析响应内容</li> <li>4（完成）响应内容解析完成，可以在客户端调用了</li></ul></li> <li>status属性：服务器（请求资源）的状态</li> <li>返回的内容 <br> <ul><li>responseText：返回以文本形式存放的内容</li> <li>responseXML：返回XML形式的内容</li></ul></li></ul></li></ul></li> </ol>
```html
<input type="button" value="按钮" id="btn" />
```
```js
var oBtn = document.getElementById('btn');
oBtn.onclick = function(){
    //打开浏览器
    /*
        1. 创建一个ajax对象，并解决兼容性问题    
    */
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    /*
    也有使用以下方式处理兼容性问题
    try {
        xhr = new XMLHttpRequest();
    } catch(e){
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    */
    //在地址栏中输入地址
    /*
        open方法：
        参数：
            1. 打开方式：用get方式还是post方式
            2. 地址
            3. 是否异步
                异步：非阻塞模式（前面的代码不会影响后面代码的执行）
                同步：阻塞模式（当前面代码没做完时，阻塞后续代码的执行）
    */
    xhr.open('get', '1.txt', true);
    //提交 发送请求
    xhr.send();
    //等待服务器返回内容
    /*
        readyState: ajax工作状态：0、1、2、3、4
        responseText: ajax请求返回的内容就被存放到这个属性下面 字符串类型
        onreadystatechange：当readyState改变的时候触发
        status：服务器状态，http状态码
    */
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                alert(xhr.responseText);
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
```
<p>try语法：</p>
```js
try {
    //代码尝试执行这个块中的内容，如果有错误，则会执行catch{}的代码，并且传入一个错误信息参数
    alert(a);
    //throw
    throw new Error('错了错了'); //即使上面没错，也可以通过throw来手动抛错
} catch (e) {
    alert(e);
}
```
<p><strong>表单</strong></p>
<ul>
<li>表单：数据的提交 <br>
<ul><li>action：数据提交的地址，默认是当前页面</li>
<li>method：数据提交的方式，默认是get方式 (不太安全，在数据传输过程中也可以修改) <br>
<ul><li>get <br>
<ul><li>把数据名称和数据值用等号连接，如果有多个的话，那么它会把多个数据组合，用&amp;进行连接。然后把数据放到url?后面传到指定页面</li>
<li>url长度限制的原因，我们不要通过get方式传递过多的数据</li>
<li>历史记录中可能会记录下来，可能会造成用户信息的暴露</li></ul></li>
<li>post <br>
<ul><li>数据串联的方式与get相同</li>
<li>数据在url后面不出现</li>
<li>理论上无长度限制</li></ul></li></ul></li>
<li>enctype: 提交的数据格式，默认是application/x-www-form-urlencoded</li></ul></li>
</ul>
<p><strong>get方式：</strong></p>
```html
<form action="1.get.php">
    <input type="text" name="username" />
    <input type="text" name="age" />
    <input type="submit" value="提交" />
</form>
```
```php
<?php
header('content-type:text/html;charset="utf-8"');
error_reporting(0);
$username = $_GET['username'];
$age = $_GET['age'];
echo "你的名字：{$username}，年龄：{$age}";
```
<p><strong>post方式：</strong></p>
```html
<form action="1.post.php" method="post">
    <input type="text" name="username" />
    <input type="text" name="age" />
    <input type="submit" value="提交" />
</form>
```
```php
<?php
header('content-type:text/html;charset="utf-8"');
error_reporting(0);
$username = $_POST['username'];
$age = $_POST['age'];
echo "你的名字：{$username}，年龄：{$age}";
```
<p><strong>get和post的区别：</strong></p>
<ul>
<li>传输方式的区别 <br>
<ul><li>get通过url地址传输</li>
<li>post通过浏览器内部传输</li></ul></li>
<li>传输数据量 <br>
<ul><li>get有数据量限制，每个浏览器都不同</li>
<li>post理论上无上限</li></ul></li>
</ul>
<p><strong>后端数据的接收：</strong></p>
<ul>
<li>$_GET <br>
<ul><li>通过URL传递给该脚本的变量的数组</li></ul></li>
<li><p>$_POST</p>

<ul><li>通过HTTP POST方法（表单）传递给该脚本的变量的数组</li></ul></li>
<li><p>前后台键名和传输方式必须一致</p></li>
<li>数据传输方式</li>
<li>数据获取方式</li>
</ul>
<h3 id="编写ajax">编写Ajax</h3>
<ul>
<li>发送请求（get和post的区别） <br>
<ul><li>send（要发送的数据）：发送请求 <br>
<ul><li>中文编码</li>
<li>缓存</li>
<li>POST: setRequestHeader(类型, 内容)：设置请求头</li>
<li>"Content-Type", "application/x-www-form-urlencoded"</li></ul></li></ul></li>
</ul>
<p><strong>应用中get和post的区别处理：</strong></p>
<p><strong>get方式传递：</strong></p>
```js
var oBtn = document.getElementById('btn');
oBtn.onclick = function(){
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    /*
        get方式的两个问题
        1. 缓存问题：在url?后面连接一个随机数或时间戳
        2. 传输中文产生乱码：采用编码encodeURI
    */
    xhr.open('get', '2.get.php?username=' + encodeURI('刘伟') + 'o&age=30&' + new Date().getTime() , true);
    xhr.send();
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                alert(xhr.responseText);
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
```
```php
<?php
header('content-type:text/html;charset="utf-8"');
error_reporting(0);
$username = $_GET['username'];
$age = $_GET['age'];
echo "你的名字：{$username}，年龄：{$age}";
```
<p><strong>post方式传递：</strong></p>
```js
var oBtn = document.getElementById('btn');
oBtn.onclick = function(){
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    xhr.open('post', '2.post.php', true);
    /*
        post方式，数据放在send()方法中，作为参数传递
        post方式要设置请求头，告诉后端发送数据的类型
        post没有缓存问题，中文也没有乱码问题，所以无需编码。
    */
    xhr.setRequestHeader('content-type', 'application/x-www.form-urlencoded'); //声明发送的数据编码类型
    xhr.send('username=刘伟&age=30');
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                alert(xhr.responseText);
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
```
```php
<?php
header('content-type:text/html;charset="utf-8"');
error_reporting(0);
$username = $_POST['username'];
$age = $_POST['age'];
echo "你的名字：{$username}，年龄：{$age}";
```
<p><strong>JSON对象</strong></p>
<p>IE7一下不支持JSON对象。可以到json.org上找javascript下面的json2.js包，引入到文件当中。</p>
<ul>
<li>stringify：可以把一个对象转成对应字符串</li>
</ul>
```js
var arr = [1, 2, 3];
var j = {left: 100};
alert(JSON.stringify(arr));
alert(JSON.stringify(j);
```
<ul>
<li>parse：可以把字符串转成对应对象</li>
</ul>
```js
var s1 = '[100, 200, 300]';
var a1 = JSON.parse(s1);
alert(a1[0]);
var s2 = '{"left": 100}'; //注意jons中的key值一定严格用双引号引起来，否则会出错
var a2 = JSON.parse(s2);
alert(a2.left);
```
<h3 id="实例ajax的封装">实例：AJAX的封装</h3>
<ul>
<li>参数 <br>
<ul><li>回调函数</li></ul></li>
<li>JSON格式的参数</li>
</ul>
```html
<input type="button" value="按钮" id="btn" />
<ul id="ul1"></ul>
```
```php
<?php
header('content-type:text/html;charset="utf-8"');
error_reporting(0);
$news = array(
    array('title'=>'新闻标题一','date'=>'2014-9-1'),
    array('title'=>'新闻标题二','date'=>'2014-9-2'),
    array('title'=>'新闻标题三','date'=>'2014-9-3'),
    array('title'=>'新闻标题四','date'=>'2014-9-4'),
    array('title'=>'新闻标题五','date'=>'2014-9-5'),
    array('title'=>'新闻标题六','date'=>'2014-9-6'),
    array('title'=>'新闻标题七','date'=>'2014-9-7'),
);
echo json_encode($news);
```
```js
oBtn.onclick = function(){
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    xhr.open('get', 'getList.php', true);
    xhr.send();
    //等待服务器返回内容
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                var data = JSON.parse(xhr.responseText);
                var oUl = document.getElementById('ul1');
                var html = '';
                for(var i=0; i<data.length; i++{
                    html += '<li><a href="">' + data[i].title + '</a>[<span>' + data[i].date + '</span>]</li>';
                }
                oUl.innerHTML = HTML;
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
```
<p><strong>为以上新闻获取添加一个定时器：</strong></p>
```js
//此处是名为ajax的js文件
function ajax(method, url, data, success){
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    if(method == 'get' && data){
        url += '?' + data;
    }
    xhr.open(method, url, true);
    if(method == 'get'){
        xhr.send();
    } else {
        xhr.setRequestHeader('content-type', 'application/x-www.form-urlencoded');
        xhr.send(data);
    }
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                success && success(xhr.responseText);
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
```
<p><strong>下面的js中引用上面的ajax函数</strong></p>
```js
oBtn.onclick = function(){
    ajax('get', 'getNews.php', '', function(data){
        var data = JSON.parse(data);
        var oUl = document.getElementById('ul1');
        var html = '';
        for(var i=0; i<data.length; i++{
            html += '<li><a href="">' + data[i].title + '</a>[<span>' + data[i].date + '</span>]</li>';
        }
        oUl.innerHTML = HTML;
    });
    setInterval(function(){
        ajax('get', 'getNews.php', '', function(data){
        var data = JSON.parse(data);
        var oUl = document.getElementById('ul1');
        var html = '';
        for(var i=0; i<data.length; i++{
            html += '<li><a href="">' + data[i].title + '</a>[<span>' + data[i].date + '</span>]</li>';
        }
        oUl.innerHTML = HTML;
    });
    }, 1000);
}
```
<h2 id="瀑布流项目开发">瀑布流项目开发</h2>
<p>瀑布流有两种展现形式：固定列、非固定列</p>
```html
<ul id="ul1">
    <li></li>
    <li></li>
    <li></li>
    <li></li>
</ul>
```
```css
body { margin: 0; }
#ul1 { width: 1080px; margin: 100px auto 0; }
li { width: 247px; list-style: none; float: left; margin-right: 10px; }
li div { border: 1px solid #000; padding: 10px; margin-bottom: 10px; }
li img { width: 225px; display: block; }
```
```php
<?php
header('Content-type:text/html; charset="utf-8"');
/*
API:
    getPics.php
        参数
        cpage：获取数据的页数
*/
$cpage = isset($_GET['cpage'])?$_GET['cpage']:1;
$url = 'http://www.wookmark.com/api/json/popular?page=' . $cpage;
$content = file_getcontents($url);
$content = iconv('gbk', 'utf-8', $content);
echo $content;
?>
```
```js
function ajax(method, url, data, success){
    var xhr = null;
    if(window.XMLHttpRequest){
        xhr = new XMLHttpRequest();
    } else {
        xhr = new ActiveXObject('Microsoft.XMLHTTP');
    }
    if(method == 'get' && data){
        url += '?' + data;
    }
    xhr.open(method, url, true);
    if(method == 'get'){
        xhr.send();
    } else {
        xhr.setRequestHeader('content-type', 'application/x-www.form-urlencoded');
        xhr.send(data);
    }
    xhr.onreadystatechange = function(){
        if(xhr.readyState == 4){
            if(xhr.status == 200){
                success && success(xhr.responseText);
            } else {
                alert('出错了，Error：' + xhr.status);
            }    
        }
    }
}
window.onload = function(){
    var oUl = document.getElementById('ul1');
    var aLi = oUl.getElementsByTagName('li');
    var iLen = aLi.length;
    var iPage = 1;
    var b = true;
    //初始化数据处理
    getList();
    function getList(){
       ajax('get', 'getPics.php', 'cpage=' + iPage, function(data){
            var data = JSON.parse(data);
            if(!data.length){
                //后续没有数据了
                return;
            }
            for (var i=0; i<data.length; i++){
                //获取高度最短的li
                var _index = getShort();
                var oDiv = document.createElement('div');
                var oImg = document.createElement('img');
                oImg.src = data[i].preview;
                oImg.style.width = '225px';
                oImg.style.height = data[i].height * （225 / data[i].width） + 'px'; //解决由于图片加载造成的图片位置添加有误的问题，并且解决图片固定宽的宽高比例问题
                oDiv.appendChild(oImg);
                var oP = document.createElement('p');
                oP.innerHTML = data[i].title;
                oDiv.appendChild(oP);
                aLi[_index].appendChild(oDiv);
            }
            b = true;
        });
    }
    window.onscroll = function(){
        var _index = getShort();
        var oLi = aLi[_index];
        var scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
        if(getTop(oLi) + oLi.offsetHeight < document.documentElement.clientHeight + scrollTop){
            if(b){
                b = false;
                iPage++;
                getList();
            }
        }
    }
    function getShort(){
        var index = 0;
        var ih = aLi[index].offsetHeight;
        for(var i=0; i<iLen; i++){
            if(aLi[i].offsetHeight < ih){
                index = i;
                ih = aLi[i].offsetHeight;
            }
        }
        return index;
    }
    function getTop(obj){
        var iTop = 0;
        while(obj){
            iTop += obj.offsetTop;
            obj = obj.offsetParent;
        }
        return iTop;
    }
}
```
<h2 id="ajax-跨域解决之jsonp">AJAX-跨域解决之JSONP</h2>
<ul>
<li>跨域的问题 <br>
<ul><li>域：域名</li>
<li>跨域请求（访问）：一个域名下的文件请求另外一个域名下的资源，就产生了跨域</li></ul></li>
<li>跨域的解决 <br>
<ul><li>Jsonp：json with padding</li></ul></li>
<li>&lt;script&gt;标签 <br>
<ul><li>src的作用：加载（包含指定的外部文件） <br>
<ul><li>可以跨域包含</li>
<li>被包含的资源可以是任何类型的文件（可以是txt、php等）</li>
<li>它只关注被包含的文件的内容是否是合法的js</li></ul></li></ul></li>
<li>原理 <br>
<ul><li>定义函数</li>
<li>包含外部文件，在被包含的文件中执行调用定义好的函数</li>
<li>参数的（数据）的实现 <br>
<ul><li>问题：包含就调用，通过动态创建&lt;script&gt;实现按需调用</li>
<li>问题：包含动态文件时可以通过一个接口实现按需生成调用函数名称</li></ul></li></ul></li>
</ul>
<p>在资源加载进来之前定义好一个函数，这个函数接收一个参数（数据），函数里面利用这么参数做一些事情。</p>
<p>然后需要的时候通过script标签加载对应远程文件资源，当远程文件资源被加载进来的时候，就会去执行我们前面定义好的函数，并且把数据当作这个函数的参数传入进去。</p>
